Mathlib/RingTheory/Valuation/Discrete/Basic.lean

```diff
@@ -369,3 +382,97 @@ theorem isUniformizer_of_maximalIdeal_eq_span [v.IsRankOneDiscrete] {r : K₀}
 end Field
 
 end Valuation
+
+namespace IsDiscreteValuationRing
+
+open IsDedekindDomain IsDedekindDomain.HeightOneSpectrum IsDiscreteValuationRing
+  IsLocalRing MonoidWithZeroHom Multiplicative Subring Valuation
+
+variable (A K : Type*) [CommRing A] [IsDomain A] [IsDiscreteValuationRing A] [Field K]
+  [Algebra A K] [IsFractionRing A K]
+
+/-- The maximal ideal of a discrete valuation ring. -/
+def maximalIdeal : HeightOneSpectrum A where
+  asIdeal := IsLocalRing.maximalIdeal A
+  isPrime := Ideal.IsMaximal.isPrime (maximalIdeal.isMaximal A)
+  ne_bot := by simpa [ne_eq, ← isField_iff_maximalIdeal_eq] using not_isField A
+
+instance isRankOneDiscrete :
+    IsRankOneDiscrete ((maximalIdeal A).valuation K) := by
+  have : Nontrivial ↥(valueGroup (valuation K (maximalIdeal A))) := by
+    let v := (maximalIdeal A).valuation K
+    let π := valuation_exists_uniformizer K (maximalIdeal A)|>.choose
+    have hπ : v π = ↑(ofAdd (-1 : ℤ)) :=
+      valuation_exists_uniformizer K (maximalIdeal A)|>.choose_spec
+    rw [Subgroup.nontrivial_iff_exists_ne_one]
+    use Units.mk0 (v π) (by simp [hπ])
+    constructor
+    · apply mem_valueGroup
+      simp only [Units.val_mk0, Set.mem_range]
+      use π
+    · simp only [hπ]
+      exact not_eq_of_beq_eq_false rfl
+  infer_instance
+
+variable {A K}
+
+open scoped WithZero
+
+theorem exists_lift_of_le_one {x : K} (H : ((maximalIdeal A).valuation K) x ≤ (1 : ℤᵐ⁰)) :
+    ∃ a : A, algebraMap A K a = x := by
+  obtain ⟨π, hπ⟩ := exists_irreducible A
+  obtain ⟨a, ⟨b, ⟨hb, h_frac⟩⟩⟩ := IsFractionRing.div_surjective (A := A) x
+  by_cases ha : a = 0
+  · rw [← h_frac]
+    use 0
+    rw [ha, map_zero, zero_div]
+  · rw [← h_frac] at H
+    obtain ⟨n, u, rfl⟩ := eq_unit_mul_pow_irreducible ha hπ
+    obtain ⟨m, w, rfl⟩ := eq_unit_mul_pow_irreducible (nonZeroDivisors.ne_zero hb) hπ
+    replace hb := (mul_mem_nonZeroDivisors.mp hb).2
+    rw [mul_comm (w : A) _, map_mul _ (u : A) _, map_mul _ _ (w : A), div_eq_mul_inv, mul_assoc,
+      Valuation.map_mul, Integers.one_of_isUnit' u.isUnit (valuation_le_one _), one_mul,
+      mul_inv, ← mul_assoc, Valuation.map_mul, map_mul, map_inv₀, map_inv₀,
+      Integers.one_of_isUnit' w.isUnit (valuation_le_one _), inv_one, mul_one, ← div_eq_mul_inv,
+      ← map_div₀, ← @IsFractionRing.mk'_mk_eq_div _ _ K _ _ _ (π ^ n) _ hb,
+      valuation_of_mk', map_pow, map_pow] at H
```

Comment:
It's annoying that we can't seem to use automation to make this line simpler but because you're rewriting `mul_assoc` and `div_eq_mul_inv` in both directions (which you need to do), I am struggling to find a trick.